### PR TITLE
[Pytorch] NVIDIA-DL-Framework-Inspect support – part 1 – core

### PR DESCRIPTION
**Pull request by pggPL**
# Description

Core part of NVIDIA-DL-Framework-Inspect support - which affects the core TE files.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/module/linear.py [hunk 1]
```diff
@@ -1106,13 +1126,29 @@ def forward(
             else:
                 bias_tensor = None
 
+            quantizers = (
+                self._get_quantizers(fp8_output, fp8_grad)
+                if not TEDebugState.debug_enabled
+                else self._get_debug_quantizers(fp8_output, fp8_grad)
+            )
+            debug = TEDebugState.debug_enabled
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/linear.py:1134]:
```suggestion
            debug = TEDebugState.debug_enabled
            quantizers = (
                self._get_quantizers(fp8_output, fp8_grad)
                if not debug
                else self._get_debug_quantizers(fp8_output, fp8_grad)
            )
```

File: transformer_engine/pytorch/module/linear.py [hunk 2]
```diff
@@ -1106,13 +1126,29 @@ def forward(
...
+            if debug:
+                if not any_feature_enabled(quantizers):
+                    # If no feature is used, then run faster implementation with debug = False.
+                    quantizers = self._get_quantizers(fp8_output, fp8_grad)
+                    debug = False
+
+            if debug and isinstance(weight_tensor, QuantizedTensor):
+                raise RuntimeError("FP8 weights are not supported in debug mode.")
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/linear.py:1142]:
```suggestion

                if isinstance(weight_tensor, QuantizedTensor):
                    raise RuntimeError("FP8 weights are not supported in debug mode.")
```

File: transformer_engine/debug/pytorch/utils.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+"""Utils functions for the debug module."""
+
+
+def any_feature_enabled(quantizers):
+    """Returns True if at least one API call is made from DebugQuantizer."""
+    for q in quantizers:
+        if q.any_feature_enabled():
+            return True
+    return False
```
Comment (ptrendx) [inline transformer_engine/debug/pytorch/utils.py:13]:
```suggestion
    return any([q.any_feature_enabled() for q in quantizers])
```

File: transformer_engine/pytorch/module/base.py [hunk 4]
```diff
@@ -1064,3 +1106,47 @@ def _load_from_state_dict(
...
+        Validate name passed to the module.
+        This is invoked in the forward() method as module names are assigned after Model is initialized in Megatron-LM.
+        If no name is assigned, it creates a default name with layer count as the variable.
+        """
+        assert TEDebugState.debug_enabled
+        import nvdlfw_inspect.api as debug_api
+
+        if self.name is None:
+            debug_api.log_message(
+                "[DEBUG-WARNING] Names are not provided to debug modules. ",
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/base.py:1121]: The tag should probably be a constant defined somewhere in debug utils rather than hardcoded here.

File: transformer_engine/pytorch/module/base.py [hunk 5]
```diff
@@ -1007,6 +1039,11 @@ def get_weight_workspace(
                     out = None
                     del self._fp8_workspaces[cache_name]
 
+            is_debug = isinstance(quantizer, DebugQuantizer)
+            is_out_debug_tensor = out is not None and isinstance(out, DebugQuantizedTensor)
+            if is_debug != is_out_debug_tensor:
+                out = None
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/base.py:1045]: Could we pass additional `debug` parameter (or just check the state in the global object) before doing those cases that could only happen during debug? I'd like to minimize the number of Python calls that we have to do.

File: transformer_engine/pytorch/module/base.py [hunk 6]
```diff
@@ -412,6 +418,9 @@ def __init__(self) -> None:
         self._fp8_workspaces: Dict[str, QuantizedTensor] = {}
         self.activation_dtype: Optional[torch.dtype] = None
 
+        if not TEDebugState.debug_enabled:
+            TEDebugState.initialize()
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/base.py:422]: I thought that `TEDebugState.debug_enabled` would be `False` when we do not want to do debug. So would we try to initialize the state every time then?

File: transformer_engine/pytorch/module/layernorm_linear.py [hunk 7]
```diff
@@ -1292,13 +1311,29 @@ def forward(
             else:
                 bias_tensor = getattr(self, self.bias_names[0])  # Unused
 
+            debug = TEDebugState.debug_enabled
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/layernorm_linear.py:1314]: Could we move this to the beginning of the function? You use the state there as well.

File: transformer_engine/pytorch/module/layernorm_mlp.py [hunk 8]
```diff
@@ -1415,17 +1473,36 @@ def forward(
             is_first_microbatch = False
 
         with self.prepare_forward(inp, num_gemms=2) as inp:
+
+            quantizers = (
+                self._get_quantizers()
+                if not TEDebugState.debug_enabled
+                else self._get_debug_quantizers()
+            )
+            debug = TEDebugState.debug_enabled
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/layernorm_mlp.py:1482]: Move to the beginning of the function.

File: transformer_engine/debug/pytorch/debug_state.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+class TEDebugState:
+    """
+    A class to manage the state of debug layers.
+    """
+
+    layer_count = 1
+    layers_initialized = {}
+    weight_tensor_tp_group_reduce = True
+    debug_enabled = None
+    initialized = False
```
Comment (ptrendx) [inline transformer_engine/debug/pytorch/debug_state.py:21]: You have the initialized variable but you don't use it - shouldn't `initialize` function check if the state was already initialized before doing anything?

File: transformer_engine/debug/pytorch/debug_quantization.py [hunk 10]
```diff
@@ -0,0 +1,524 @@
...
+            args["tensor"] = columnwise_gemm_tensor
+            args["rowwise"] = False
+            debug_api.transformer_engine.inspect_tensor_postquantize(**args)
+
+    def quantize(
+        self,
+        tensor: torch.Tensor,
+        *,
+        out: Optional[Union[torch.Tensor, DebugQuantizedTensor]] = None,
+        dtype: torch.dtype = None,
```
Comment (ptrendx) [inline transformer_engine/debug/pytorch/debug_quantization.py:254]: What is the use case for this?

File: transformer_engine/pytorch/module/base.py (hunk 4 above)
Comment (pggPL) [reply transformer_engine/pytorch/module/base.py:1121]: I removed this prefix for now, since we do not use it in the other places.

File: transformer_engine/pytorch/module/base.py (hunk 5 above)
Comment (pggPL) [reply transformer_engine/pytorch/module/base.py:1045]: We will need to have logic of changing quantizers there in the near future - when quantizer does not match the cache tensor. I can leave this for now and this will need to be changed in the PR which would enable recipe change during training.

File: transformer_engine/debug/pytorch/debug_quantization.py (hunk 10 above)
Comment (pggPL) [reply transformer_engine/debug/pytorch/debug_quantization.py:254]: Weight caching for fake quantization for example. We want to save them in correct precision.

File: transformer_engine/pytorch/module/base.py (hunk 6 above)
Comment (pggPL) [reply transformer_engine/pytorch/module/base.py:422]: to raise an error if one module is initialized before usage of nvinspect api, and another is initialized after

File: transformer_engine/debug/pytorch/debug_state.py (hunk 9 above)
Comment (pggPL) [reply transformer_engine/debug/pytorch/debug_state.py:21]: Oh, I haven't removed it and lint does not error it for some reason. Now we have 3 states of debug enabled true/false/none indicating that. Maybe it would result is some small speedup during initialization, but I don't think it is worth it.

File: transformer_engine/pytorch/utils.py [hunk 11]
```diff
@@ -329,6 +330,16 @@ def round_up_to_nearest_multiple(value, multiple):
...
 
 
+def needs_quantized_gemm(obj, rowwise=True):
+    """Used to check if obj will need quantized gemm or normal gemm."""
+    if isinstance(obj, DebugQuantizedTensor):
+        return (
+            type(obj.get_tensor(not rowwise))  # pylint: disable=unidiomatic-typecheck
+            is not torch.Tensor
+        )
+    return type(obj) is not torch.Tensor  # pylint: disable=unidiomatic-typecheck
```
Comment (ksivaman) [inline transformer_engine/pytorch/utils.py:340]:
```suggestion
        return not isinstance(obj.get_tensor(not rowwise), torch.Tensor)
    return not isinstance(obj, torch.Tensor)
```

File: transformer_engine/pytorch/module/base.py [hunk 12]
```diff
@@ -1042,6 +1079,11 @@ def get_weight_workspace(
             else:
                 tex.quantize(tensor, quantizer, out, skip_update_flag)
 
+        if not needs_quantized_gemm(type(out)):  # only holds for debug quantizer
```
Comment (ksivaman) [inline transformer_engine/pytorch/module/base.py:1082]:
```suggestion
        if not needs_quantized_gemm(out):  # only holds for debug quantizer
```

File: transformer_engine/pytorch/cpp_extensions/gemm.py [hunk 13]
```diff
@@ -109,6 +110,21 @@ def general_gemm(
...
+        A = A.get_tensor(not transa)
+        B = B.get_tensor(transb)
+
+        assert (type(A) in [torch.Tensor, torch.nn.parameter.Parameter]) == (
+            type(B) in [torch.Tensor, torch.nn.parameter.Parameter]
+        ), (
+            "[Debug tools] Processed tensors should both be FP8 tensors or both be torch tensors  "
+            f"            but type(A) = {type(A)},             "
+            f"  type(B) = {type(B)}"
+        )
```
Comment (ksivaman) [inline transformer_engine/pytorch/cpp_extensions/gemm.py:126]: This feels like an awkward check, could you explain it, i.e. what are we trying to verify here? Can `get_tensor` even return non `torch.Tensor` types? Note that `torch.nn.parameter.Parameter` is `torch.Tensor` as well, so if we are checking this, it's better to move to the suggested change.

Comment (ksivaman) [inline transformer_engine/pytorch/cpp_extensions/gemm.py:126]:
```suggestion
        assert isinstance(A, torch.Tensor) == isinstance(B, torch.Tensor), (
            "[Debug tools] Processed tensors should both be FP8 tensors or both be torch tensors  "
            f"            but type(A) = {type(A)},             "
            f"  type(B) = {type(B)}"
        )
```

File: transformer_engine/pytorch/utils.py (hunk 11 above)
Comment (ptrendx) [reply transformer_engine/pytorch/utils.py:340]: @ksivaman I don't believe this suggestion is correct. Basically you want to differentiate here between regular torch.Tensor and quantized tensor (which could be a subclass of torch.Tensor). In my internal quantizer PR #1638 I introduce a function to check if it is quantized tensor, maybe that would be better instead.

File: transformer_engine/pytorch/cpp_extensions/gemm.py (hunk 13 above)
Comment (ptrendx) [reply transformer_engine/pytorch/cpp_extensions/gemm.py:126]: QuantizedTensor is a subclass of torch.Tensor so it would pass yield `True` in `isinstance(A, torch.Tensor)`. See my previous comment.

Comment (pggPL) [reply transformer_engine/pytorch/cpp_extensions/gemm.py:126]: It was checking if both tensors are torch.Tensor ot Float8Tensors. Since more tensor class were added in the meantime and this check is not necessary, I removed it.

File: transformer_engine/debug/pytorch/debug_quantization.py (hunk 10 above)
Comment (pggPL) [reply transformer_engine/debug/pytorch/debug_quantization.py:254]: Suppose weight is stored in FP32, but activation is in BF16. Then we want weight after fake-quantization to be in BF16 to prevent cast each forward.

File: transformer_engine/debug/pytorch/debug_state.py [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+            cls.debug_enabled = debug_api.DEBUG_MANAGER is not None
+
+    @classmethod
+    def reset(cls):
+        """Resets layer count and stats buffers."""
+        from ..features.utils.stats_buffer import STATS_BUFFERS
+
+        STATS_BUFFERS.reset()
+        cls.debug_enabled = None
+        cls.layers_initialized.clear()
```
[the remark this answers is not in the document]
Comment (pggPL) [reply transformer_engine/debug/pytorch/debug_state.py:49]: this call is used mostly for testing - your scenario would work

File: transformer_engine/pytorch/module/base.py [hunk 15]
```diff
@@ -1042,6 +1079,11 @@ def get_weight_workspace(
             else:
                 tex.quantize(tensor, quantizer, out, skip_update_flag)
 
+        if not needs_quantized_gemm(out):  # only holds for debug quantizer
+            assert (
+                out.dtype == workspace_dtype
+            ), "Activation dtype cannot be changed with nvidia-dlframework-inspect."
```
[the remark this answers is not in the document]
Comment (pggPL) [reply transformer_engine/pytorch/module/base.py]: I moved this check to the debug quantizer function, when this is trivial what does error mean.

File: transformer_engine/debug/pytorch/debug_state.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (ptrendx) [reply transformer_engine/debug/pytorch/debug_state.py:49]: Then maybe let's change it to `_reset`?

[the remark this answers is not in the document]
Comment (pggPL) [reply transformer_engine/debug/pytorch/debug_state.py:49]: done

File: transformer_engine/pytorch/module/linear.py [hunk 16]
```diff
@@ -1193,8 +1230,9 @@ def forward(
 
     def _get_quantizers(self, fp8_output, fp8_grad):
         if not self.fp8:
-            return [None] * 5
+            return [None] * 6
         grad_input_quantizer = None
+        grad_weight_quantizer = None
```
Comment (timmoon10) [inline transformer_engine/pytorch/module/linear.py:1235]: I strongly oppose adding a quantizer for the grad weight. It doesn't make sense logically since we never plan to output a quantized grad weight, barring some major development in optimizers. I get that the asymmetry is weird, but the real problem is that `output_quantizer` and `grad_input_quantizer` are also illogical. We should be using the `input_quantizer`/`grad_output_quantizer` from the layer that consumes them, like what te.Sequential does, but that's very tricky to implement.

**Review by timmoon10 — CHANGES_REQUESTED** [review]

File: transformer_engine/pytorch/module/linear.py (hunk 16 above)
Comment (pggPL) [reply transformer_engine/pytorch/module/linear.py:1235]: Ok, but maybe one wants to do some experiments with weight quantization and debug tools helps with that. Moreover this quantizer enables us to log the statistics.

Comment (ptrendx) [reply transformer_engine/pytorch/module/linear.py:1235]: @timmoon10 Actually, I would argue that `grad_weight_quantizer` makes more sense compared to the other quantizers you listed since it is actually "local" vs the other ones logically belonging to the other operations. Even in te.Sequential the grad_weight quantizer would belong to the op that holds those weights.
I also don't agree that `we never plan to output a quantized grad weight` - while the convergence story of this would need to be understood, projects like MS-AMP tried to provide FP8 gradient functionality since it saves a lot of memory during training. So it is not completely out of the question that we will want to provide that as well.

File: transformer_engine/pytorch/module/linear.py [hunk 17]
```diff
@@ -840,6 +851,7 @@ def __init__(
         ub_bulk_dgrad: bool = False,
         ub_bulk_wgrad: bool = False,
         ub_name: Optional[str] = None,
+        name: Optional[str] = None,
```
[the remark this answers is not in the document]
Comment (pggPL) [reply transformer_engine/pytorch/module/linear.py:887]: I added some simple docs and I will use it in the NVTX markers and add it in error messages in the another PR.

File: transformer_engine/debug/pytorch/debug_quantization.py (hunk 10 above)
Comment (pggPL) [reply transformer_engine/debug/pytorch/debug_quantization.py:254]: we do it here https://github.com/NVIDIA/TransformerEngine/blob/66d6afbf6e4bccd3d40e85a3d5ed307229137304/transformer_engine/pytorch/module/linear.py#L200

**Review by ptrendx — APPROVED** [review]
From my side LGTM.

File: transformer_engine/pytorch/module/linear.py (hunk 16 above)
Comment (timmoon10) [reply transformer_engine/pytorch/module/linear.py:1268]: The MS-AMP case is a good example. I suppose the right answer would be to gracefully support multiple quantizer types. In most cases we would have `grad_weight_quantizer=None`. `grad_weight_quantizer` would only be non-trivial if we want quantized grads or we're doing something weird like debug mode.

Unresolving for future visibility.

**Review by timmoon10 — APPROVED** [review]
I haven't fully gone through the code, but things look reasonable to me. I would be fine merging.
